### ISSUE TITLE
[QST] cudaErrorMemoryAllocation when running tests

### ISSUE THREAD

**Issue by c-jamie** [rapidsai/cudf#6237]
**Background**

I've successfullly built cuda + tests on my local machine, however for a large number of tests, I'm getting `cudaErrorMemoryAllocation` errors

**What is your question**

1) is this expected behaviour for my system; if it because my graphics hardware is insufficient (env is below), is there a recommended graphics card memory size?
2) is there any way of getting these tests to run on my current system (without buying new graphics hardware) - I'm specifically interested in `JOIN_TEST` - if I could get this to run on it's own, that's enough

**Build command**

```
INSTALL_PREFIX=${INSTALL_PREFIX:=${PREFIX:=${CONDA_PREFIX}}}

PARALLEL_LEVEL=4 cmake -DCMAKE_INSTALL_PREFIX=${INSTALL_PREFIX} \
      -DCMAKE_CXX11_ABI=ON \
      -DGPU_ARCHS=75 \
      -DUSE_NVTX=ON \
      -DCMAKE_CUDA_ARCHITECTURES=75\
      -DDISABLE_DEPRECATION_WARNING=ON \
      -DPER_THREAD_DEFAULT_STREAM=OFF \
      -DCMAKE_BUILD_TYPE=Debug \
      -GNinja \
      ..
```

**Example test run error**

```
(cudf) user@home:~/code/cudf/cpp/build/gtests$ ./JOIN_TEST
terminate called after throwing an instance of 'rmm::bad_alloc'
  what():  std::bad_alloc: CUDA error at: /home/user/miniconda/envs/cudf/include/rmm/mr/device/cuda_memory_resource.hpp:68: cudaErrorMemoryAllocation out of memory
Aborted (core dumped)
```

**env**

<details><summary>Click here to see environment details</summary><pre>

     **git***
     commit 1a3b3f217be93a55b47af3a9d0da29f0fcb7c7e9 (HEAD -> branch-0.15, origin/branch-0.15, origin/HEAD)
     Merge: a5dda7faf cddd339be
     Author: Keith Kraus <kkraus@nvidia.com>
     Date:   Fri Aug 21 14:30:28 2020 -0400

     Merge pull request #6057 from nvdbaranec/column_reorder_fix

     [REVIEW] Fix for reading columns out-of-order they appear in parquet files.
     **git submodules***

     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=18.04
     DISTRIB_CODENAME=bionic
     DISTRIB_DESCRIPTION="Ubuntu 18.04.5 LTS"
     NAME="Ubuntu"
     VERSION="18.04.5 LTS (Bionic Beaver)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 18.04.5 LTS"
     VERSION_ID="18.04"
     HOME_URL="https://www.ubuntu.com/"
     SUPPORT_URL="https://help.ubuntu.com/"
     BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
     PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
     VERSION_CODENAME=bionic
     UBUNTU_CODENAME=bionic
     Linux cillairne 5.4.0-47-generic #51~18.04.1-Ubuntu SMP Sat Sep 5 14:35:50 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

     ***GPU Information***
     Tue Sep 15 16:34:51 2020
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 450.51.06    Driver Version: 450.51.06    CUDA Version: 11.0     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |                               |                      |               MIG M. |
     |===============================+======================+======================|
     |   0  Quadro K2200        On   | 00000000:02:00.0 Off |                  N/A |
     | 42%   34C    P8     1W /  39W |    132MiB /  4034MiB |      0%      Default |
     |                               |                      |                  N/A |
     +-------------------------------+----------------------+----------------------+
     |   1  GeForce GTX 165...  On   | 00000000:03:00.0 Off |                  N/A |
     | 35%   44C    P8     6W / 100W |      6MiB /  3911MiB |      0%      Default |
     |                               |                      |                  N/A |
     +-------------------------------+----------------------+----------------------+

     +-----------------------------------------------------------------------------+
     | Processes:                                                                  |
     |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
     |        ID   ID                                                   Usage      |
     |=============================================================================|
     |    0   N/A  N/A      1319      G   /usr/lib/xorg/Xorg                 56MiB |
     |    0   N/A  N/A      1527      G   /usr/bin/gnome-shell               72MiB |
     |    1   N/A  N/A      1319      G   /usr/lib/xorg/Xorg                  4MiB |
     +-----------------------------------------------------------------------------+

     ***CPU***
     Architecture:        x86_64
     CPU op-mode(s):      32-bit, 64-bit
     Byte Order:          Little Endian
     CPU(s):              24
     On-line CPU(s) list: 0-23
     Thread(s) per core:  2
     Core(s) per socket:  12
     Socket(s):           1
     NUMA node(s):        1
     Vendor ID:           GenuineIntel
     CPU family:          6
     Model:               63
     Model name:          Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz
     Stepping:            2
     CPU MHz:             1198.726
     CPU max MHz:         3300.0000
     CPU min MHz:         1200.0000
     BogoMIPS:            4987.87
     Virtualisation:      VT-x
     L1d cache:           32K
     L1i cache:           32K
     L2 cache:            256K
     L3 cache:            30720K
     NUMA node0 CPU(s):   0-23
     Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti intel_ppin ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm xsaveopt cqm_llc cqm_occup_llc dtherm ida arat pln pts md_clear flush_l1d

     ***CMake***

     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
     Copyright (C) 2017 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

     ***nvcc***
     /usr/local/cuda/bin/nvcc
     nvcc: NVIDIA (R) Cuda compiler driver
     Copyright (c) 2005-2019 NVIDIA Corporation
     Built on Sun_Jul_28_19:07:16_PDT_2019
     Cuda compilation tools, release 10.1, V10.1.243

     ***Python***
     /home/cillairne/miniconda/bin/python
     Python 3.7.7

     ***Environment Variables***
     PATH                            : /usr/local/cuda/bin:/home/cillairne/miniconda/bin:/home/cillairne/miniconda/condabin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
     LD_LIBRARY_PATH                 : /usr/local/cuda/lib64:/usr/local/cuda/lib64
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    : /home/cillairne/miniconda
     PYTHON_PATH                     :

     ***conda packages***
     /home/cillairne/miniconda/bin/conda
     # packages in environment at /home/cillairne/miniconda:
     #
     # Name                    Version                   Build  Channel
     _libgcc_mutex             0.1                        main    defaults
     alsa-lib                  1.2.3                h516909a_0    conda-forge
     brotlipy                  0.7.0           py37h7b6447c_1000    defaults
     ca-certificates           2020.6.20            hecda079_0    conda-forge
     cairo                     1.16.0            h3fc0475_1005    conda-forge
     certifi                   2020.6.20        py37hc8dfbb8_0    conda-forge
     cffi                      1.14.0           py37he30daa8_1    defaults
     chardet                   3.0.4                 py37_1003    defaults
     conda                     4.8.3                    py37_0    defaults
     conda-package-handling    1.6.1            py37h7b6447c_0    defaults
     cryptography              2.9.2            py37h1ba5d50_0    defaults
     fontconfig                2.13.1            h1056068_1002    conda-forge
     freetype                  2.10.2               he06d7ca_0    conda-forge
     giflib                    5.2.1                h516909a_2    conda-forge
     glib                      2.65.0               h3eb4bd4_0    defaults
     gmock                     1.10.0                        4    conda-forge
     graphite2                 1.3.13            he1b5a44_1001    conda-forge
     harfbuzz                  2.7.1                hee91db6_0    conda-forge
     icu                       67.1                 he1b5a44_0    conda-forge
     idna                      2.10                       py_0    defaults
     jpeg                      9d                   h516909a_0    conda-forge
     lcms2                     2.11                 hbd6801e_0    conda-forge
     ld_impl_linux-64          2.33.1               h53a641e_7    defaults
     libedit                   3.1.20191231         h7b6447c_0    defaults
     libffi                    3.3                  he6710b0_2    defaults
     libgcc-ng                 9.1.0                hdf63c60_0    defaults
     libiconv                  1.16                 h516909a_0    conda-forge
     libpng                    1.6.37               hed695b0_2    conda-forge
     libstdcxx-ng              9.1.0                hdf63c60_0    defaults
     libtiff                   4.1.0                hc7e4089_6    conda-forge
     libuuid                   2.32.1            h14c3975_1000    conda-forge
     libwebp-base              1.1.0                h516909a_3    conda-forge
     libxcb                    1.13              h14c3975_1002    conda-forge
     libxml2                   2.9.10               h68273f3_2    conda-forge
     lz4-c                     1.9.2                he1b5a44_3    conda-forge
     ncurses                   6.2                  he6710b0_1    defaults
     openjdk                   11.0.8               hacce0ff_0    conda-forge
     openssl                   1.1.1g               h516909a_1    conda-forge
     pcre                      8.44                 he1b5a44_0    conda-forge
     pip                       20.1.1                   py37_1    defaults
     pixman                    0.38.0            h516909a_1003    conda-forge
     pthread-stubs             0.4               h14c3975_1001    conda-forge
     pycosat                   0.6.3            py37h7b6447c_0    defaults
     pycparser                 2.20                       py_0    defaults
     pyopenssl                 19.1.0                   py37_0    defaults
     pysocks                   1.7.1                    py37_0    defaults
     python                    3.7.7                hcff3b4d_5    defaults
     python_abi                3.7                     1_cp37m    conda-forge
     readline                  8.0                  h7b6447c_0    defaults
     requests                  2.24.0                     py_0    defaults
     ruamel_yaml               0.15.87          py37h7b6447c_1    defaults
     setuptools                47.3.1                   py37_0    defaults
     six                       1.15.0                     py_0    defaults
     sqlite                    3.32.3               h62c20be_0    defaults
     tk                        8.6.10               hbc83047_0    defaults
     tqdm                      4.47.0                     py_0    defaults
     urllib3                   1.25.9                     py_0    defaults
     wheel                     0.34.2                   py37_0    defaults
     xorg-fixesproto           5.0               h14c3975_1002    conda-forge
     xorg-inputproto           2.3.2             h14c3975_1002    conda-forge
     xorg-kbproto              1.0.7             h14c3975_1002    conda-forge
     xorg-libice               1.0.10               h516909a_0    conda-forge
     xorg-libsm                1.2.3             h84519dc_1000    conda-forge
     xorg-libx11               1.6.12               h516909a_0    conda-forge
     xorg-libxau               1.0.9                h14c3975_0    conda-forge
     xorg-libxdmcp             1.1.3                h516909a_0    conda-forge
     xorg-libxext              1.3.4                h516909a_0    conda-forge
     xorg-libxfixes            5.0.3             h516909a_1004    conda-forge
     xorg-libxi                1.7.10               h516909a_0    conda-forge
     xorg-libxrender           0.9.10            h516909a_1002    conda-forge
     xorg-libxtst              1.2.3             h516909a_1002    conda-forge
     xorg-recordproto          1.14.2            h516909a_1002    conda-forge
     xorg-renderproto          0.11.1            h14c3975_1002    conda-forge
     xorg-xextproto            7.3.0             h14c3975_1002    conda-forge
     xorg-xproto               7.0.31            h14c3975_1007    conda-forge
     xz                        5.2.5                h7b6447c_0    defaults
     yaml                      0.2.5                h7b6447c_0    defaults
     zlib                      1.2.11               h7b6447c_3    defaults
     zstd                      1.4.5                h6597ccf_2    conda-forge

</pre></details>

**Comment by kkraus14** [comment]
@c-jamie it looks like GPU 0 in your machine is a Quadro K2200 which is Maxwell generation and is unsupported. I assume you meant to target GPU 1 which is your GTX 1650? Could you try doing `CUDA_VISIBLE_DEVICES=1 ./JOIN_TEST`?

**Comment by c-jamie** [comment]
@kkraus14 thanks for the response. Strange when I target device 1, I get `cudaErrorNoKernelImageForDevice`. I'm building with  `compute_75` which is what I though the correct arch was for my GTX. Have I made a mistake?

**Comment by kkraus14** [comment]
Hmm it's very strange you're getting a `cudaErrorInvalidDevice` error. Are you running in a container or somewhere else that isolates the GPU you're targetting already? The `nvidia-smi` report is showing a K2200 and GTX 1650.

It's possible our join test uses more than 4GB of memory. cc @jrhemstad who I think would know

**Comment by c-jamie** [comment]
@kkraus14 I've just removed the K2200 from my box to remove ambiguity. Now `./HASH_MAP_TEST` runs fine, but still `cudaErrorMemoryAllocation` on `./COLUMN_TEST` and `./JOIN_TEST`.

**Comment by davidwendt** [comment]
@c-jamie Just out of curiosity, could you try running these with `--rmm_mode=cuda`?
Example:
```
./COLUMN_TEST --rmm_mode=cuda
```

**Comment by c-jamie** [comment]
@davidwendt that change works, thank you very much. Is there anywhere I can go to read documentation on these modes and their meaning?

**Comment by davidwendt** [comment]
The modes are set in the gtests through the command line. The available modes for gtests:
https://github.com/rapidsai/cudf/blob/b35882895513143bb05e7906c525f9f83788fea9/cpp/tests/utilities/base_fixture.hpp#L268-L275
There is some documentation here: https://github.com/rapidsai/rmm#available-resources

The default is `pool` mode which pre-allocates half the GPU memory and then grows from there as needed. I could be that these two tests are hitting an edge case which makes the pool grow too large. The `cuda` mode does not use a pool and only allocates and frees as needed for each test.
@harrism

**Comment by c-jamie** [comment]
Got it thanks. This can be closed.

**Comment by harrism** [comment]
I think this needs a bit of investigation. It is strange that it would OOM with the pool but not with the cuda allocator.  I wonder if the test only OOMs after running many other tests (causing fragmentation). I would be surprised, though, because that would imply a memory leak somewhere (otherwise all fragments should be coalesced between tests).

**Comment by harrism** [comment]
I just tried running with the initial pool size set to 2GiB and the max pool size set to 4 GiB to try to simulate what must be happening on this GTX 1650. In the RMM debug log, I never see the available pool memory dropping below 2147400000 bytes.

So JOIN_TEST is not using a lot of memory. Not sure why it would fail on GTX 1650, but I don't have one to test on.
